hiro/cocoa/widget/hex-edit.cpp

```diff
@@ -5,43 +5,167 @@
 }
 
-@end
-
-namespace hiro {
-
-auto pHexEdit::construct() -> void {
-  cocoaView = cocoaHexEdit = [[CocoaHexEdit alloc] initWith:self()];
+- (NSTableView *) tableView {
+  return tableView;
 }
 
-auto pHexEdit::destruct() -> void {
-  [cocoaView removeFromSuperview];
+- (hiro::mHexEdit *) hexEdit {
+  return hexEdit;
 }
 
-auto pHexEdit::setAddress(u32 offset) -> void {
-}
 
-auto pHexEdit::setBackgroundColor(Color color) -> void {
+- (NSInteger) numberOfRowsInTableView:(NSTableView *)tableView {
+  return (max(1, hexEdit->length()) + hexEdit->columns() - 1) / hexEdit->columns();
 }
 
-auto pHexEdit::setColumns(u32 columns) -> void {
-}
-
-auto pHexEdit::setForegroundColor(Color color) -> void {
+-(id) tableView:(NSTableView *) tableView
+objectValueForTableColumn:(NSTableColumn *) tableColumn
+            row:(NSInteger) row {
+  // return a string with the content of (row, column)
+  // u32 address = hexEdit->address() + row * hexEdit->columns();
+  u32 address = row * hexEdit->columns();
+  // address only
+  if ([[tableColumn identifier] isEqualToString:@"Address"]) {
+    return [NSString stringWithUTF8String:hex(address, 8L).data()];
+  } else if ([[tableColumn identifier] isEqualToString:@"Char"]){
+    // create the char string by concatenating
+    NSMutableString *output = [NSMutableString stringWithCapacity:hexEdit->columns()];
+    for (auto column : range(hexEdit->columns())) {
+      if (address < hexEdit->length()) {
+        u8 data = hexEdit->doRead(address++);
+        [output appendString:[NSString stringWithFormat:@"%c",
+                              (data >= 0x20 && data <= 0x7e ? (char)data : '.')]];
+      }
+    }
+    return output;
+  } else {
+    // return only the single byte at the correct position
+    NSInteger columnNumber = [tableColumn.identifier integerValue];
+    address += columnNumber;
+    if (address < hexEdit->length()) {
+      u8 data = hexEdit->doRead(address);
+      return [NSString stringWithUTF8String:hex(data, 2L).data()];
+    } else {
+      return @"  ";
+    }
+  }
 }
 
-auto pHexEdit::setLength(u32 length) -> void {
+-(void) tableView:(NSTableView *) tableView
+   setObjectValue:(id) object
+   forTableColumn:(NSTableColumn *) tableColumn
+              row:(NSInteger) row {
+  // when table is edited, modify underlying data source
+  NSInteger colNumber = [tableColumn.identifier integerValue];
+  u32 address = row * hexEdit->columns() + colNumber;
+  
+  // unclear if this is needed, but just to make sure the cast is safe
+  if ([object isKindOfClass:[NSString class]]) {
+    if (address < hexEdit->length()) {
+      // only get the first 2 characters
+      NSString* newVal = [(NSString *)object substringToIndex:2];
+      NSScanner* hexScanner = [NSScanner scannerWithString:newVal];
+      unsigned int data = 0;
+      if ([hexScanner scanHexInt:&data]) {
+        // this.......is probably ok....???
+        hexEdit->doWrite(address, (u8)data);
+      }
+    }
+    
+  }
```

Comment:
I don't think this check is needed. The only way we end up here is when the user enters text into an `NSTextFieldCell`, so the value should in all circumstances already be coerced to a string:

```suggestion
  if (address < hexEdit->length()) {
    // only get the first 2 characters
    NSString* newVal = [(NSString *)object substringToIndex:2];
    NSScanner* hexScanner = [NSScanner scannerWithString:newVal];
    unsigned int data = 0;
    if ([hexScanner scanHexInt:&data]) {
      // this.......is probably ok....???
      hexEdit->doWrite(address, (u8)data);
    }
  }
```